src/metal.jl

```diff
@@ -332,6 +334,78 @@ function add_address_spaces!(@nospecialize(job::CompilerJob), mod::LLVM.Module,
     return new_f
 end
 
+# add addrspace 2 to global constants
+function add_global_address_spaces!(@nospecialize(job::CompilerJob), mod::LLVM.Module)
+    for gv in globals(mod)
+        if isconstant(gv) && addrspace(value_type(gv)) == 0
+            gv_ty = global_value_type(gv)
+            gv_name = LLVM.name(gv)
+            
+            new_gv = GlobalVariable(mod, gv_ty, "", 2)
+            
+            alignment!(new_gv, alignment(gv))
+            unnamed_addr!(new_gv, unnamed_addr(gv))
+            initializer!(new_gv, initializer(gv))
+            constant!(new_gv, true)
+            linkage!(new_gv, linkage(gv))
+            visibility!(new_gv, visibility(gv))
+
+            funcs = Set{LLVM.Function}()
+            for use in uses(gv)
+                inst = user(use)
+                bb = LLVM.parent(inst)
+                f = LLVM.parent(bb)
+
+                push!(funcs, f)
+            end
+
+            for f in funcs
+                ft = function_type(f)
+                new_f = LLVM.Function(mod, "h", ft)
```

Comment:
Isn't it overkill to use a value mapper here? Can't you just position an IRBuilder before the old uses and emit new IR + RAUW + delete?

Reply:
Is there a simple way in LLVM.jl to replace uses in a given function?

Reply:
Why do you only want to replace uses within a given function?

Reply:
because I am using a hack which stores the global in an alloca at the beginning of the function, so that I'dont need to update all bitcasts and geps which would need their ptr operand to be updated to ptr addrspace(2).

Reply:
> similar to how we handle byref args

That we do because it is generally annoying to replace the IR dealing with pointers to IR dealing with values. Updating the IR to simply deal with a pointer that's in a different address space shouldn't be so problematic, no? I can take a look later if you want.

Reply:
I think you'd need alias analysis to make this work. Since there could be ptr to int and the pointer may escape somehow. Feel free to give it a shot though.

Reply:
That also applies to your current implementation, no? If a GV in AS 0 gets converted to an integer and later loaded from again, the current version will only change the AS of the pointer flowing into the `inttoptr`, so you'd end up with a load from the original (wrong) AS too.

Reply:
Anyway, turns out LLVM's Clone utils don't "recursively" update uses of a GV that changed address spaces (the initial GEP still returned an untyped pointer, failing IR verification). So I kept your general idea, but managed to get rid of the `alloca` by using a ConstExpr. Will push in a sec.

Reply:
LGTM